File: 599-minimum-index-sum-of-two-lists/solution.cpp
class Solution {
public:
    vector<string> findRestaurant(vector<string>& list1, vector<string>& list2) {
        map<string, int> hashMap;
        for ( int i = 0; i < list1.size(); i ++ ) {
            hashMap.insert(pair<string, int>(list1[i], i));
        }
        int best = 1e9, tmp = 0;
        vector<string> result;
        for ( int j = 0; j < list2.size(); j ++) {
            if ( 0 == hashMap.count(list2[j]) ) {
                tmp = 1e9;
            } else {
                tmp = hashMap[list2[j]];
            }
            if ( j + tmp < best ) {
                vector<string> res;
                res.push_back(list2[j]);
                result = res;
                best = j + tmp;
            } else if ( j + tmp == best ) {
                result.push_back(list2[j]);
            }
        }
        return result;
    }
};

File: 599-minimum-index-sum-of-two-lists/solution.py
class Solution:
    def findRestaurant(self, list1, list2):
        """
        :type list1: List[str]
        :type list2: List[str]
        :rtype: List[str]
        """
        hash_map = {k: idx for idx, k in enumerate(list1)}
        best, result = 1e9, []
        for idx, k in enumerate(list2):
            if hash_map.get(k, 1e9) + idx < best:
                best = hash_map.get(k, 1e9) + idx
                result = [k]
            elif hash_map.get(k, 1e9) + idx == best:
                result.append(k)
        return result